src/libcollections/btree/node.rs

```diff
@@ -1221,6 +1231,155 @@ impl<K, V> Node<K, V> {
     }
 }
 
+// Helper struct for storing information about one level of the tree.
+struct Level<L, U> {
+    elements_remaining: usize,
+    nodes_remaining: usize,
+    cur_node: Option<Node<L, U>>,
+    cur_node_remaining: usize,
+}
+
+impl<L, U> Level<L, U> {
+    fn new_node(&mut self, node: Node<L, U>, capacity: usize, minimum: usize) {
+        // Make the new node the current node for this level.
+        mem::replace(&mut self.cur_node, Some(node));
+
+        // Determine how many elements the new node should hold, so that no node is overfull
+        // or underfull. This tries to produce as many full nodes as possible.
+        self.cur_node_remaining = if self.nodes_remaining == 1 {
+            self.elements_remaining
+        } else {
+            cmp::min(capacity, self.elements_remaining - minimum)
+        };
+        self.nodes_remaining -= 1;
+    }
+
+    // Return mutable reference to current node. Creates a new leaf node if necessary.
+    fn get_or_create_leaf(&mut self, capacity: usize, minimum: usize) -> &mut Node<L, U> {
+        if self.cur_node.is_none() {
+            self.new_node(Node::new_leaf(capacity), capacity, minimum);
+        }
+
+        // `unwrap` won't panic because we just inserted a new node if there wasn't one
+        // before.
+        self.cur_node.as_mut().unwrap()
+    }
+
+    // Return mutable reference to current node. Creates a new internal node if necessary.
+    fn get_or_create_internal(&mut self, capacity: usize, minimum: usize) -> &mut Node<L, U> {
+        if self.cur_node.is_none() {
+            // `unsafe` is needed for `new_internal`.
+            unsafe {
+                self.new_node(Node::new_internal(capacity), capacity, minimum);
+            }
+        }
+
+        // `unwrap` won't panic because we just inserted a new node if there wasn't one
+        // before.
+        self.cur_node.as_mut().unwrap()
+    }
+}
+
+impl<K, V> Node<K, V> {
+    pub fn from_sorted_iter<I>(iter: I, len: usize, b: usize) -> (usize, Node<K, V>)
+            where I: Iterator<Item=(K, V)> {
+        // Compute level structure
+        let capacity = capacity_from_b(b);
+        let minimum = min_load_from_capacity(capacity);
+        let mut num_elements = len;
+        // Allocate memory for the worst-case number of levels.
+        // let max_height = (0.5*((num_elements + 1) as f64)).log(minimum as f64) as usize;
+        // let mut levels = Vec::with_capacity(max_height);
```

Comment:
Yes, the idea was to minimize the number of allocations. But this isn't possible with the changes I've pushed just yet.